### PR TITLE
[Skills& VA] Fallback rule support

### LINKED ISSUE #1598: Remove "do you want to hear more" type prompts from Skills
## Project
Skills
C#

## Description
Skill flow update: Remove the “do you want to hear more” type prompt after every productivity interaction

## To Reproduce

Using calendar skill to get summary of todays' meeting, by end of the conversation, a prompt ask user "Do you want to hear the content of first event" and user has to answer yes or no.

## Expected behavior

Change the prompt to informational msg and allow user to switch to other dialogs.

### PR DESCRIPTION
**Pull request by DingmaomaoBJTU**
### Purpose
[Part 2] Close #1598 Remove "do you want to hear more" type prompts from Skills

### Changes
1. Add fallback event send & handle in Email/Calendar skill, tested
2. Add fallback event redispatch handle logic in VA

### Tests
Menual tested all possible senarios.

### Feature Plan
*Are there any remaining steps or dependencies before this issue can be fully resolved? If so, describe and link to any relevant pull requests or issues.*

#### General
- [x] I have commented my code, particularly in hard-to-understand areas	
- [ ] I have added or updated the appropriate tests	
- [ ] I have updated related documentation

#### Bots
- [x] I have validated that new and updated responses use appropriate [Speak](https://docs.microsoft.com/en-us/azure/bot-service/dotnet/bot-builder-dotnet-text-to-speech?view=azure-bot-service-3.0) and [InputHint](https://docs.microsoft.com/en-us/azure/bot-service/dotnet/bot-builder-dotnet-add-input-hints?view=azure-bot-service-3.0) properties to ensure a high-quality speech-first experience
- [ ] I have replicated language model changes across the English, French, Italian, German, Spanish, and Chinese `.lu` files and validated that deployment is successful

#### Deployment Scripts
- [ ] I have replicated my changes in the **Virtual Assistant Template** and **Sample** projects
- [ ] I have replicated my changes in the **Skill Template** and **Sample** projects

### CODE REVIEW

File: templates/Virtual-Assistant-Template/csharp/Sample/VirtualAssistantSample/Dialogs/MainDialog.cs [hunk 1]
```diff
@@ -22,7 +24,7 @@
 
 namespace VirtualAssistantSample.Dialogs
 {
-    public class MainDialog : RouterDialog
+    public class MainDialog : DispatchDialog
```
Comment (lauren-mills) [inline templates/Virtual-Assistant-Template/csharp/Sample/VirtualAssistantSample/Dialogs/MainDialog.cs]: Why did you change the name? Seems like you could just add the method you need to the routerDialog. #Resolved

File: templates/Virtual-Assistant-Template/csharp/Sample/VirtualAssistantSample/Dialogs/MainDialog.cs [hunk 2]
```diff
@@ -184,6 +186,27 @@ protected override async Task RouteAsync(DialogContext dc, CancellationToken can
             }
         }
 
+        protected override async Task RedispatchAsync(DialogContext innerDc, DialogTurnResult result = null, CancellationToken cancellationToken = default(CancellationToken))
```
Comment (lauren-mills) [inline templates/Virtual-Assistant-Template/csharp/Sample/VirtualAssistantSample/Dialogs/MainDialog.cs]: Why do you need a separate method here? Essentially this is doing the same thing as RouteAsync #Resolved

File: templates/Virtual-Assistant-Template/csharp/Sample/VirtualAssistantSample/Services/SkillIntentRecognizer.cs [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+        private BotSettings _settings;
+        private bool _confirmIntentSwitch = true;
+
+        public SkillIntentRecognizer(BotServices services, BotSettings settings)
+        {
+            _services = services;
+            _settings = settings;
+        }
+
+        public async Task<string> RecognizeSkillIntentAsyncFunc(DialogContext dc)
```
Comment (lauren-mills) [inline templates/Virtual-Assistant-Template/csharp/Sample/VirtualAssistantSample/Services/SkillIntentRecognizer.cs:24]: I also don't understand why this is needed. The IsSkill method already does this logic, so you're only adding the confirmIntentSwitch setting which could come from appsettings #ByDesign

File: templates/Virtual-Assistant-Template/csharp/Sample/VirtualAssistantSample/VirtualAssistantSample.csproj [hunk 4]
```diff
@@ -37,6 +35,11 @@
     <PackageReference Include="Microsoft.VisualStudio.Web.CodeGeneration.Design" Version="2.2.0" />
   </ItemGroup>
 
+  <ItemGroup>
```
Comment (lauren-mills) [inline templates/Virtual-Assistant-Template/csharp/Sample/VirtualAssistantSample/VirtualAssistantSample.csproj]: These projects should not be directly referenced in the sample or template. #WontFix

**Review by lauren-mills — CHANGES_REQUESTED** [review]

File: templates/Virtual-Assistant-Template/csharp/Sample/VirtualAssistantSample/Dialogs/MainDialog.cs (hunk 1 above)
Comment (DingmaomaoBJTU) [reply templates/Virtual-Assistant-Template/csharp/Sample/VirtualAssistantSample/Dialogs/MainDialog.cs]: Simplify it in lib pr -- [https://github.com/microsoft/botframework-solutions/pull/1976](https://github.com/microsoft/botframework-solutions/pull/1976)

---
In reply to: [308303059](https://github.com/microsoft/botframework-solutions/pull/1971#discussion_r308303059) [](ancestors = 308303059)

File: templates/Virtual-Assistant-Template/csharp/Sample/VirtualAssistantSample/Dialogs/MainDialog.cs (hunk 2 above)
Comment (DingmaomaoBJTU) [reply templates/Virtual-Assistant-Template/csharp/Sample/VirtualAssistantSample/Dialogs/MainDialog.cs]: Removed

---
In reply to: [308303486](https://github.com/microsoft/botframework-solutions/pull/1971#discussion_r308303486) [](ancestors = 308303486)

File: templates/Virtual-Assistant-Template/csharp/Sample/VirtualAssistantSample/VirtualAssistantSample.csproj (hunk 4 above)
Comment (DingmaomaoBJTU) [reply templates/Virtual-Assistant-Template/csharp/Sample/VirtualAssistantSample/VirtualAssistantSample.csproj]: As it has dependency with lib changes, I had to reference them directly as future branch has no pipeline planed. will remove this when lib change picked up to next branch and built into preview pkg.

---
In reply to: [308304732](https://github.com/microsoft/botframework-solutions/pull/1971#discussion_r308304732) [](ancestors = 308304732)

File: templates/Virtual-Assistant-Template/csharp/Sample/VirtualAssistantSample/Services/SkillIntentRecognizer.cs (hunk 3 above)
Comment (DingmaomaoBJTU) [reply templates/Virtual-Assistant-Template/csharp/Sample/VirtualAssistantSample/Services/SkillIntentRecognizer.cs:24]: This is the interface called by skilldialog which has no way to get cognitive models in VA.

---
In reply to: [308304362](https://github.com/microsoft/botframework-solutions/pull/1971#discussion_r308304362) [](ancestors = 308304362)

**Review by lauren-mills — APPROVED** [review]

File: lib/csharp/Directory.Build.props [hunk 5]
```diff
@@ -8,7 +8,7 @@
   </PropertyGroup>
 
   <PropertyGroup>
-    <TreatWarningsAsErrors>True</TreatWarningsAsErrors>
+    <TreatWarningsAsErrors>False</TreatWarningsAsErrors>
```
Comment (lzc850612) [inline lib/csharp/Directory.Build.props]: revert this. #Resolved

File: skills/src/csharp/Skills.sln [hunk 6]
```diff
@@ -37,6 +37,10 @@ Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "WeatherSkill", "experimenta
 EndProject
 Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "BingSearchSkill", "experimental\bingsearchskill\bingsearchskill\BingSearchSkill.csproj", "{8CB101FA-B496-4507-80A2-151FAB7E8ED1}"
 EndProject
+Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "Microsoft.Bot.Builder.Skills", "..\..\..\lib\csharp\microsoft.bot.builder.skills\Microsoft.Bot.Builder.Skills\Microsoft.Bot.Builder.Skills.csproj", "{B25E0652-3BCE-4EEE-86CB-F38660CC9736}"
+EndProject
+Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "Microsoft.Bot.Builder.Solutions", "..\..\..\lib\csharp\microsoft.bot.builder.solutions\microsoft.bot.builder.solutions\Microsoft.Bot.Builder.Solutions.csproj", "{2BBFC7CC-34F2-4D05-9DDE-99380E2E592E}"
```
Comment (lzc850612) [inline skills/src/csharp/Skills.sln]: we have a daily build in myget now. it's in https://botbuilder.myget.org/feed/aitemplates/package/nuget/Microsoft.Bot.Builder.Skills #Resolved

File: skills/src/csharp/calendarskill/calendarskill/CalendarSkill.csproj [hunk 7]
```diff
@@ -195,6 +195,11 @@
     <EmbeddedResource Include="Utilities\WindowsIanaMapping" />
   </ItemGroup>
 
+  <ItemGroup>
+    <ProjectReference Include="..\..\..\..\..\lib\csharp\microsoft.bot.builder.skills\Microsoft.Bot.Builder.Skills\Microsoft.Bot.Builder.Skills.csproj" />
+    <ProjectReference Include="..\..\..\..\..\lib\csharp\microsoft.bot.builder.solutions\microsoft.bot.builder.solutions\Microsoft.Bot.Builder.Solutions.csproj" />
```
Comment (lzc850612) [inline skills/src/csharp/calendarskill/calendarskill/CalendarSkill.csproj]: use the myget feed. #Resolved

File: skills/src/csharp/emailskill/emailskill/EmailSkill.csproj [hunk 8]
```diff
@@ -179,6 +179,11 @@
     <EmbeddedResource Include="Responses\ShowEmail\ShowEmailResponses.json" />
     <EmbeddedResource Include="Responses\ShowEmail\ShowEmailResponses.zh.json" />
   </ItemGroup>
+
+  <ItemGroup>
+    <ProjectReference Include="..\..\..\..\..\lib\csharp\microsoft.bot.builder.skills\Microsoft.Bot.Builder.Skills\Microsoft.Bot.Builder.Skills.csproj" />
+    <ProjectReference Include="..\..\..\..\..\lib\csharp\microsoft.bot.builder.solutions\microsoft.bot.builder.solutions\Microsoft.Bot.Builder.Solutions.csproj" />
```
Comment (lzc850612) [inline skills/src/csharp/emailskill/emailskill/EmailSkill.csproj]: same #Resolved

File: templates/Virtual-Assistant-Template/csharp/VirtualAssistantTemplate.sln [hunk 9]
```diff
@@ -1,7 +1,7 @@
 ﻿
 Microsoft Visual Studio Solution File, Format Version 12.00
-# Visual Studio Version 16
-VisualStudioVersion = 16.0.29009.5
+# Visual Studio 15
```
Comment (lzc850612) [inline templates/Virtual-Assistant-Template/csharp/VirtualAssistantTemplate.sln]: undo this. #Resolved

File: templates/Virtual-Assistant-Template/csharp/VirtualAssistantTemplate.sln [hunk 10]
```diff
@@ -13,6 +13,10 @@ Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "VirtualAssistantSample.Test
 EndProject
 Project("{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}") = "VirtualAssistantVSIX", "VSIX\VirtualAssistantVSIX.csproj", "{794F4E29-D65D-4F5F-9B04-55FFD5EE864D}"
 EndProject
+Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "Microsoft.Bot.Builder.Skills", "..\..\..\lib\csharp\microsoft.bot.builder.skills\Microsoft.Bot.Builder.Skills\Microsoft.Bot.Builder.Skills.csproj", "{8807EF9A-9974-4E7F-ADBD-B9E916039155}"
+EndProject
```
Comment (lzc850612) [inline templates/Virtual-Assistant-Template/csharp/VirtualAssistantTemplate.sln]: same #Resolved

**Review by lzc850612 — CHANGES_REQUESTED** [review]

File: skills/src/csharp/Skills.sln (hunk 6 above)
Comment (DingmaomaoBJTU) [reply skills/src/csharp/Skills.sln]: Upgrade to preview lib~

---
In reply to: [310825398](https://github.com/microsoft/botframework-solutions/pull/1971#discussion_r310825398) [](ancestors = 310825398)

**Review by lzc850612 — APPROVED** [review]

**Review by KayMKM — APPROVED** [review]

**Pull request merged**